sympy/sets/conditionset.py

```diff
@@ -190,11 +195,33 @@ def ok_sig(a, b):
                 ok_sig(i, j) for i, j in zip(a, b))
         if not ok_sig(self.sym, other):
             return S.false
+        # if either argument of And is False,
+        # we must return False even if we
+        # get a TypeError on the other
+
+        # try doing base_cond first and return
+        # False immediately if it is False
         try:
-            return And(
-                Contains(other, self.base_set),
-                Lambda((self.sym,), self.condition)(other))
+            base_cond = Contains(other, self.base_set)
+            base_error = False
         except TypeError:
+            base_cond = None
+            base_error = True
+        if (not base_error) and (base_cond is S.false):
+            return False
+
+        try:
+            lambda_cond = (Lambda((self.sym,), self.condition)(other)).doit()
+            lambda_error = False
+        except TypeError:
+            lambda_cond = None
+            lambda_error = True
+        if (not lambda_error) and (lambda_cond is S.false):
+            return False
```

Comment:
I find this code quite convoluted and I think I prefer the way it was written before. This would be clearer I think without the `lambda_cond` and `lambda_error` variables.

I don't like catching `TypeError` as it makes me think that either we should be checking the types before the operation or the operation shouldn't raise a `TypeError`. Catching the `TypeError` seems like working around either a bug or a poor definition of what the scope of the function is.

I'm also not sure that we should call doit here as that can lead to arbitrarily expensive evaluation.

Reply:
> I find this code quite convoluted and I think I prefer the way it was written before. This would be clearer I think without the lambda_cond and lambda_error variables.

The reason why it was done is to allow one or more of [these](https://github.com/sympy/sympy/issues/20101#issuecomment-694959855) problems to be fixed. (I can't remember which ones).

> I don't like catching TypeError as it makes me think that either we should be checking the types before the operation or the operation shouldn't raise a TypeError.

Reason for doing so is [here](https://github.com/sympy/sympy/pull/20121#discussion_r491949054).

> I'm also not sure that we should call doit here as that can lead to arbitrarily expensive evaluation.

This was needed so that we don't get errors for the following added tests:
https://github.com/sympy/sympy/pull/20121/files#diff-91900d1e8fe044ea6c7601b777d5f895R38-R44
This matrix functionality will be used in the stats module soon. Current master should not raise with these new tests if we replace `... in ...` with `Contains(..., ...)`. But I think `in` is more user-friendly than `Contains`.

Reply:
> But I think `in` is more user-friendly than `Contains`.

We shouldn't use `in` in the codebase because there is no general way to stop it from raising. Python requires it to give True or False (unlike e.g. `<`) and many queries are indeterminate e.g.:
```julia
In [1]: x = Symbol('x')                                                                                                                        

In [2]: x in Reals                                                                                                                             
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-c53dfd933383> in <module>
----> 1 x in Reals

~/current/sympy/sympy/sympy/sets/sets.py in __contains__(self, other)
    673             # x in y must evaluate to T or F; to entertain a None
    674             # result with Set use y.contains(x)
--> 675             raise TypeError('did not evaluate to a bool: %r' % c)
    676         return b
    677 

TypeError: did not evaluate to a bool: (-oo < x) & (x < oo)
```

> the `doit()` call should not be expensive since 99.99% of the time, it's calling doit on `S.true` or `S.false`.

Well then it's not needed 99.99% of the time :)

We should only use `doit` in library code if we're sure we know what sympy classes it is going to act on. If we add a `doit` in `_contains` then there is no way for a user to prevent it in the situations where it is slow.

Reply:
> Well then it's not needed 99.99% of the time :)

:smirk: That's a dangerous argument. Division by 0 checks aren't needed often either...

I agree with your other arguments though and I'll remove the `doit()`.